### PR TITLE
Upgrade to ramsey/uuid:^3.0

### PR DESCRIPTION
**Pull request by Lctrs**

### CODE REVIEW

**Comment by Lctrs** [comment]
I see 2 ways to prevent BC Break and support both ramsey/uuid 2.x and 3.x

1) Use https://github.com/MattKetmo/uuid-2x-bridge

Main drawback is that it tells composer it replaces ramsey/uuid 2.x and uses 3.x behind the scenes. So it still requires us to bump our minimal required version to PHP 5.4.

2) Write our own Bridge that will check for existence of `Rhumsaa\Uuid\Uuid` or `Ramsey\Uuid\Uuid` and map our functions calls to the existing class.

Main drawback of this is typehint in IDE I think.

Which one do you prefer ?

**Comment by xabbuh** [comment]
@Lctrs My idea is the second one. Mainly, introduce a `Uuid` class in the `Model` namespace which basically consists only of `fromString()` factory method. Under the hood, it would then use the `Uuid` class from the other package with the namespace depending on their presence. We would still need to draft a new major release, but we could support both versions of the UUID library at the same time. Do you want to make these changes?

**Comment by Lctrs** [comment]
Here you go. WDYT @xabbuh ?

File: .travis.yml [hunk 1]
```diff
@@ -20,7 +19,7 @@ php:
 matrix:
   fast_finish: true
   include:
-    - php: 5.3
+    - php: 5.4
```
Comment (xabbuh) [inline .travis.yml]: Let's revert the changes for the PHP version right now. We can consider to do that later.

File: spec/StatementIdSpec.php [hunk 2]
```diff
@@ -12,8 +12,9 @@
 namespace spec\Xabbuh\XApi\Model;
 
 use PhpSpec\ObjectBehavior;
-use Rhumsaa\Uuid\Uuid;
+use spec\Xabbuh\XApi\Model\Uuid\WrongUuid;
```
Comment (xabbuh) [inline spec/StatementIdSpec.php]: I would define this class in this file right after the spec class (we do not need it anywhere else).

Comment (xabbuh) [reply spec/StatementIdSpec.php]: Oh wait, I think we don't need this at all (see my comments below).

File: spec/Uuid/WrongUuid.php [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace spec\Xabbuh\XApi\Model\Uuid;
+
+class WrongUuid
+{
+}
```
Comment (xabbuh) [inline spec/Uuid/WrongUuid.php]: missing trailing newline

File: src/StatementId.php [hunk 4]
```diff
@@ -29,8 +27,12 @@ private function __construct()
     {
     }
 
-    public static function fromUuid(Uuid $uuid)
+    public static function fromUuid($uuid)
```
Comment (xabbuh) [inline src/StatementId.php]: I would keep the type-hint, but use our own `Uuid` class instead.

Comment (xabbuh) [reply src/StatementId.php]: Oh, and that would have to documented in the changelog file.

File: src/Uuid.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?php
+
```
Comment (xabbuh) [inline src/Uuid.php:2]: Please add the header block (as present in other source files).

File: src/Uuid.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+namespace Xabbuh\XApi\Model;
+
+final class Uuid
```
Comment (xabbuh) [inline src/Uuid.php:20]: Please add yourself as the author of this class.

File: src/Uuid.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    /**
+     * @param string $uuid
+     * @return \Rhumsaa\Uuid\Uuid|\Ramsey\Uuid\Uuid
+     */
+    public static function fromString($uuid)
+    {
+        if (class_exists('Rhumsaa\Uuid\Uuid')) {
+            return \Rhumsaa\Uuid\Uuid::fromString($uuid);
```
Comment (xabbuh) [inline src/Uuid.php]: This method should return an instance of our own `Uuid` class. This way, no information is leaked to the outside about which UUID implementation we use internally.

Comment (xabbuh) [reply src/Uuid.php]: I would also prefer to add `use` statements for both classes (we can alias them as `RhumsaaUuid` and `RamseyUuid`).

**Comment by xabbuh** [comment]
@Lctrs I just left some minor comments. But overall the changes look pretty good. 👍

**Comment by Lctrs** [comment]
Should be good now.

For the changelog, in which version should I put my change ?

File: src/Uuid.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    public static function fromString($uuid)
+    {
+        if (class_exists('Rhumsaa\Uuid\Uuid')) {
+            return new self(RhumsaaUuid::fromString($uuid));
+        }
+
+        return new self(RamseyUuid::fromString($uuid));
+    }
+
+    public function toString()
```
Comment (xabbuh) [inline src/Uuid.php]: I think we can use `__toString()` here (none of the other model classes uses `toString()`, but use the magic method instead).

File: src/Uuid.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return $this->uuid->toString();
+    }
+
+    public function equals(Uuid $uuid)
+    {
+        return $this->uuid->toString() === $uuid->toString();
```
Comment (xabbuh) [inline src/Uuid.php]: Then we could here just compare the string representation of both instances (`(string) $this === (string) $uuid`).

**Comment by xabbuh** [comment]
Since types used in type hints will change, we need to release a new major version. Thus, you can document the change under the new `2.0.0` headline. Could you then also update the branch alias in the `composer.json` file accordingly?

**Comment by Lctrs** [comment]
Hi @xabbuh. Happy new year !

Currently with the new `Uuid` class, it's harder to create instance of this class with new uuid (in previous version, we could do `$statementId = StatementId::fromUuid(\Rhumsaa\Uuid\Uuid::uuid4());` for example which is not possible with the new class).

I think I'll introduce uuid1, uuid2, uuid3, uuid4 functions to the Uuid class to keep this feature. WDYT ?

**Comment by xabbuh** [comment]
@Lctrs Happy new year for you too! 🎉 

Wouldn't it be a simpler solution to make the constructor of the new `Uuid` class `public` which would then generate a v4 UUID? If I do not miss anything, there is no need to use UUIDs following any of the other versions.

**Comment by Lctrs** [comment]
Users of MySQL may want to generate UUID v1 in combination with `OrderedTimeCodec` to optimize UUIDs storing in the database.

Not my case but this need could exists.

**Comment by xabbuh** [comment]
It looks like the spec does not require a specific UUID version. So having factory methods for each of them inside the new `Uuid` class looks like the best idea to me.

**Comment by Lctrs** [comment]
Last commits pushed.

Before releasing 2.0, don't forget to drop deprecated functions. :)

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
Thank you @Lctrs. I have merged your changes into the `master` branch in 55cba18809a6f3cd7f4417a7fb7d369a82083cbe.

**Pull request closed without merging**